### ISSUE TITLE
`ffi.gc` not working with circular references?

### ISSUE THREAD

**Issue by nicolasnoble** [LuaJIT/LuaJIT#896]
Consider this very synthetic example for the sake of demonstrating the problem:

```lua
local ffi = require('ffi')

ffi.cdef [[
void* malloc(size_t);
void free(void*);
]]

function test1(name)
  local o = { name = name }
  local mt = { __gc = function() print(o.name) end }
  o.proxy = newproxy()
  debug.setmetatable(o.proxy, mt)
  return o
end

function test2(name)
  local o = { name = name }
  o.proxy = ffi.gc(ffi.C.malloc(1), function() print(o.name) ffi.C.free(o.proxy) end)
  return o
end

test1('Using newproxy (1)')
print 'Collecting garbage...'
collectgarbage()
test1('Using newproxy (2)')
print 'Collecting garbage...'
collectgarbage()

test2('Using ffi.gc (1)')
print 'Collecting garbage...'
collectgarbage()
test2('Using ffi.gc (2)')
print 'Collecting garbage...'
collectgarbage()

print 'Ending...'
```

On a fresh build of LuaJIT with no special parameter on my Ubuntu x86_64, I get the following output:

```
$ luajit ./testme.lua
Collecting garbage...
Using newproxy (1)
Collecting garbage...
Using newproxy (2)
Collecting garbage...
Collecting garbage...
Ending...
Using ffi.gc (2)
Using ffi.gc (1)
```

The expectation would be that both codepaths would work in a similar way. Testing shows that the references held with `ffi.gc` will never get garbage collected, until the Lua state is actually closed. Signs point to the circular reference to `o` inside of `ffi.gc`, because this works as expected:

```lua
local ffi = require('ffi')

ffi.cdef [[
void* malloc(size_t);
void free(void*);
]]

function test3(name)
  local o = { name = name }
  o.proxy = ffi.gc(ffi.C.malloc(1), function(p) print('removing ' .. tostring(p)) ffi.C.free(p) end)
  return o
end

test3('Using ffi.gc (1)')
print 'Collecting garbage...'
collectgarbage()
test3('Using ffi.gc (2)')
print 'Collecting garbage...'
collectgarbage()

print 'Ending...'
```
```
$ ./luajit testme2.lua
Collecting garbage...
removing cdata<void *>: 0x55fa57729480
Collecting garbage...
removing cdata<void *>: 0x55fa57729480
Ending...
```

Note that the circular reference may only be adjacent to `ffi.gc` to also cause the problem:

```lua
local ffi = require('ffi')

ffi.cdef [[
void* malloc(size_t);
void free(void*);
]]

function test4(name)
  local o = { name = name }
  o.proxy = ffi.gc(ffi.C.malloc(1), function(p) print(o.name) ffi.C.free(p) end)
  return o
end

test4('Using ffi.gc (1)')
print 'Collecting garbage...'
collectgarbage()
test4('Using ffi.gc (2)')
print 'Collecting garbage...'
collectgarbage()

print 'Ending...'
```
```
Collecting garbage...
Collecting garbage...
Ending...
Using ffi.gc (1)
Using ffi.gc (2)
```

**Comment by MikePall** [comment]
The point of `ffi.gc()` is to provide a simple finalization mechanism for **external** FFI data. Second-guessing the garbage collector with an internal Lua proxy object and a separate closure for each use of `ffi.gc()` is a bad idea. Finalizers are already very, very expensive.
There's no way to solve this with the current GC, since the FFI finalizer table would have to be an ephemeron table. Which is horribly expensive (quadratic), especially if you have many objects.
In other words: you're trying too hard to solve the wrong issue. Think outside the box about ways to get by without this complexity.
